src/queue.rs

```diff
@@ -126,43 +135,51 @@ impl<M: GuestAddressSpace> DescriptorChain<M> {
     }
 
     fn is_valid(&self) -> bool {
-        !(self
-            .mem
-            .checked_offset(self.addr, self.len as usize)
-            .is_none()
-            || (self.has_next() && self.next >= self.queue_size))
+        self.mem
+            .checked_offset(self.desc.addr(), self.desc.len as usize)
+            .filter(|_| !self.has_next() || self.desc.next < self.queue_size)
+            .is_some()
     }
 
     /// Checks if this descriptor chain has another descriptor chain linked after it.
     pub fn has_next(&self) -> bool {
-        self.flags & VIRTQ_DESC_F_NEXT != 0 && self.ttl > 1
+        self.desc.flags & VIRTQ_DESC_F_NEXT != 0 && self.ttl > 1
     }
 
-    /// Checks if the driver designated this as a write only descriptor.
-    ///
-    /// If this is false, this descriptor is read only.
-    /// Write only means the the emulated device can write and the driver can read.
-    pub fn is_write_only(&self) -> bool {
-        self.flags & VIRTQ_DESC_F_WRITE != 0
+    /// Return a `GuestMemory` object that can be used to access the buffers
+    /// pointed to by the descriptor chain.
+    pub fn memory(&self) -> &M::M {
+        &*self.mem
     }
+}
+
+impl<M: GuestAddressSpace> Iterator for DescriptorChain<M> {
+    type Item = Descriptor;
 
     /// Returns the next descriptor in this descriptor chain, if there is one.
     ///
     /// Note that this is distinct from the next descriptor chain returned by
     /// [`AvailIter`](struct.AvailIter.html), which is the head of the next
     /// _available_ descriptor chain.
-    pub fn next_descriptor(&self) -> Option<DescriptorChain<M>> {
-        if self.has_next() {
-            Self::read_new(
-                self.mem.clone(),
-                self.desc_table,
-                self.ttl - 1,
-                self.queue_size,
-                self.next,
-            )
-        } else {
-            None
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.ttl == 0 {
+            return None;
         }
+
+        let curr = self.desc;
+        if !self.has_next() {
+            self.ttl = 0
+        } else {
+            self.index = self.desc.next;
+            let desc_table_size = size_of::<Descriptor>() * self.queue_size as usize;
+            let slice = self.mem.get_slice(self.desc_table, desc_table_size).ok()?;
+            self.desc = slice
+                .get_array_ref(0, self.queue_size as usize)
+                .ok()?
+                .load(self.index as usize);
+            self.ttl = self.ttl - 1;
+        }
```

Comment:
Nitpicking, but clippy complains about this:
```
warning: manual implementation of an assign operation
   --> src/queue.rs:180:13
    |
180 |             self.ttl = self.ttl - 1;
    |             ^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `self.ttl -= 1`
    |
    = note: `#[warn(clippy::assign_op_pattern)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assign_op_pattern
```

Reply:
Fixed.